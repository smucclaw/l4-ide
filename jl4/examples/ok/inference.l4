ยง `Type inference example(s)`

inc n MEANS n + 1

singleton MEANS 5 FOLLOWED BY EMPTY

five MEANS
  CONSIDER singleton
  WHEN EMPTY THEN 0
  WHEN x FOLLOWED BY xs THEN x

six MEANS inc five

sum list MEANS
  CONSIDER list
  WHEN EMPTY THEN 0
  WHEN x FOLLOWED BY xs THEN x + sum xs

and list MEANS
  CONSIDER list
  WHEN EMPTY THEN TRUE
  WHEN x FOLLOWED BY xs THEN x AND and xs

#CHECK inc
#CHECK singleton
#CHECK five
#CHECK six
#CHECK sum
#CHECK and
