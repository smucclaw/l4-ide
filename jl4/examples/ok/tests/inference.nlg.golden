`inc`
`singleton`
`five`
`six`
`sum`
`and`